### PR TITLE
mgr/dashboard: grafana panels for rgw multisite sync performance

### PR DESCRIPTION
**Pull request by alfonsomthd**
* RGW sync perf. counters are now exposed through grafana panels.
* Sync Performance tab is only shown if rgw realm is detected.
* Prometheus module: added metrics suitable for prometheus consumption (from existing ones, not replacing for backward compatibility: it is to be backported, a follow-up cleanup PR can be done). 

Fixes: https://tracker.ceph.com/issues/45310
Signed-off-by: Alfonso Martínez <almartin@redhat.com>

HOW TO TEST IT:
1) 1st cluster: configure multisite: create realm, master zonegroup & master zone.
At least 1 daemon should be running on master zone.
https://docs.ceph.com/docs/master/radosgw/multisite/#requirements-and-assumptions
2) 2nd cluster: pull the realm  and create a secondary zone.
At least a daemon should be running on secondary zone.
3) Operate on this secondary zone (create buckets, add medium-sized objects)
and check that 'Sync Performance' graphs are reflecting this activity.
You can also use this benchmarking tool (thanks @cbodley for pointing it out):
https://github.com/markhpc/hsbench

45310-grafana-rgw-sync-info-4-panels

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by alfonsomthd** [comment]
jenkins test dashboard backend

**Comment by alfonsomthd** [comment]
@cbodley @mattbenjamin Can you take a brief look to ensure that counters/metrics info are exposed properly?

**Comment by alfonsomthd** [comment]
jenkins test make check

**Comment by alfonsomthd** [comment]
jenkins test dashboard

**Comment by callithea** [comment]
jenkins test dashboard backend

**Comment by cbodley** [comment]
> @cbodley @mattbenjamin Can you take a brief look to ensure that counters/metrics info is exposed properly?

Thanks @alfonsomthd. We had a discussion last week about different ways to dimension this data - for example, by zone rather than per gateway. In terms of the graphs themselves, can we choose different units?

For "Avg. object bytes replicated", units of KB/MB/GB will be more natural. Ideally we'd be able to represent this as throughput (ie MB/sec).. do you know if that's possible?

For "Avg. latency", I would recommend starting with milliseconds.

**Comment by mattbenjamin** [comment]
What is the y axis on the "avg objects already replicated" graph(s)--and what does it represent?  I suspect tracking the rate of replication in objects/s would be desirable.  We have an ability to see how fast objects are being synced not considering size, do we not?

**Comment by cbodley** [comment]
> What is the y axis on the "avg objects already replicated" graph(s)--and what does it represent?

this corresponds to the "fetch not modified" counter, which is incremented each time a replication request returns 304 NotModified. this means our current copy of the object is at least as recent as the source zone's

>  I suspect tracking the rate of replication in objects/s would be desirable. We have an ability to see how fast objects are being synced not considering size, do we not?

the object count is tracked in the "avgcount" part of "fetch bytes":
```
        "fetch bytes": {
            "avgcount": 59,
            "sum": 151501
        },
```
i hope we can find a way to display this as objects/sec rather than an object count that increases over time

**Comment by alfonsomthd** [comment]
> i hope we can find a way to display this as objects/sec rather than an object count that increases over time
Done. Now there is a label displaying "Objects/s".

**Comment by alfonsomthd** [comment]
> > @cbodley @mattbenjamin Can you take a brief look to ensure that counters/metrics info is exposed properly?
> 
> Thanks @alfonsomthd. We had a discussion last week about different ways to dimension this data - for example, by zone rather than per gateway. In terms of the graphs themselves, can we choose different units?
> 
> For "Avg. object bytes replicated", units of KB/MB/GB will be more natural. Ideally we'd be able to represent this as throughput (ie MB/sec).. do you know if that's possible?
> 
> For "Avg. latency", I would recommend starting with milliseconds.

Done.

**Comment by alfonsomthd** [comment]
> Looks good @alfonsomthd ! Perhaps using stacked graphs to make them easier to read? And is it possible to make the legend scroll bars hide?

Done.

**Comment by alfonsomthd** [comment]
@cbodley @mattbenjamin @epuertat your comments have been addressed.
In the description there is a screenshot reflecting it.

**Comment by alfonsomthd** [comment]
@LenzGr @theanalyst Can you give it a try and share your thoughts?

**Comment by callithea** [comment]
- [pkg build](https://shaman.ceph.com/builds/ceph/wip-laura-testing-34863/) [WIP]
- QA run [PENDING]

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-daemon-list/rgw-daemon-list.component.spec.ts [hunk 1]
```diff
@@ -29,12 +36,42 @@ describe('RgwDaemonListComponent', () => {
...
+    getPermissionsSpy.and.returnValue(new Permissions({}));
+    getRealmsSpy = spyOn(TestBed.get(RgwSiteService), 'get');
+    getRealmsSpy.and.returnValue(of([]));
     fixture = TestBed.createComponent(RgwDaemonListComponent);
     component = fixture.componentInstance;
-    fixture.detectChanges();
   });
 
   it('should create', () => {
     expect(component).toBeTruthy();
```
Comment (Devp00l) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-daemon-list/rgw-daemon-list.component.spec.ts:48]: Could you please add `fixture.detectChanges();` before this line to initialize the component correctly?

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-daemon-list/rgw-daemon-list.component.spec.ts [hunk 2]
```diff
@@ -29,12 +36,42 @@ describe('RgwDaemonListComponent', () => {
...
   });
+
+  it('should only show Daemons List tab', () => {
+    fixture.detectChanges();
+
+    const tabs = fixture.debugElement.nativeElement.querySelectorAll('tab');
+    expect(tabs.length).toEqual(1);
+    expect(tabs[0].getAttribute('heading')).toEqual('Daemons List');
```
Comment (Devp00l) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-daemon-list/rgw-daemon-list.component.spec.ts:56]: How about putting this part into a function like

```ts
  const expectTabsAndHeading = (length: number, heading: string) => {
    const tabs = fixture.debugElement.nativeElement.querySelectorAll('tab');
    expect(tabs.length).toEqual(length);
    expect(tabs[0].getAttribute('heading')).toEqual(heading);
 }
```

File: src/pybind/mgr/dashboard/tests/test_rgw_client.py [hunk 3]
```diff
@@ -124,6 +124,28 @@ def test_get_placement_targets_from_realm_zone(self, zonegroup_map, zone_info):
...
+        ]
+
+        instance = RgwClient.admin_instance()
+        expected_result = [
+            "realm1",
+            "realm2"
+                ]
+        self.assertEqual(expected_result, instance.get_realms())
+        expected_result = []
+        self.assertEqual(expected_result, instance.get_realms())
```
Comment (Devp00l) [inline src/pybind/mgr/dashboard/tests/test_rgw_client.py:147]: As expected_results is assigned 2 different lists that are only used once how about this:

```suggestion
        self.assertEqual([ "realm1", "realm2" ], instance.get_realms())
        self.assertEqual([], instance.get_realms())
```

**Comment by callithea** [comment]
jenkins test dashboard backend

**Comment by alfonsomthd** [comment]
@Devp00l Adressed all your comments.

**Comment by callithea** [comment]
- new [pkg build](https://shaman.ceph.com/builds/ceph/wip-laura-testing-34863-2/) [WIP]
- QA run [PENDING]

**Comment by epuertat** [comment]
@alfonsomthd my hints on the titles:
- Titles in Proper Case (like this)
- Remove "Avg." (assumed implicitly)
- "Replication (throughput) from Source Zone" and - "Replication (objects) from Source Zone"

**Comment by callithea** [comment]
jenkins test dashboard backend

File: src/pybind/mgr/dashboard/tests/test_rgw_client.py (hunk 3 above)
Comment (alfonsomthd) [reply src/pybind/mgr/dashboard/tests/test_rgw_client.py]: Done.

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-daemon-list/rgw-daemon-list.component.spec.ts (hunk 2 above)
Comment (alfonsomthd) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-daemon-list/rgw-daemon-list.component.spec.ts]: Done.

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-daemon-list/rgw-daemon-list.component.spec.ts (hunk 1 above)
Comment (alfonsomthd) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-daemon-list/rgw-daemon-list.component.spec.ts:55]: Done.

**Comment by alfonsomthd** [comment]
> @alfonsomthd my hints on the titles:
> 
>     * Titles in Proper Case (like this)
> 
>     * Remove "Avg." (assumed implicitly)
> 
>     * "Replication (throughput) from Source Zone" and - "Replication (objects) from Source Zone"

Done.

**Pull request merged**